### ISSUE TITLE
Bug in transformer configuration file?

### ISSUE THREAD

**Issue by zinechant** [ARM-software/SCALE-Sim#29]
First I believe the configuration assumes the sentence length is 1 (in #words) for both encoder and decoder.

I believe the following corresponds to one head of an attention layer (either self or cross attention).
Layer name, IFMAP Height, IFMAP Width, Filter Height, Filter Width, Channels, Num Filter, Strides.
Attn_Xmap,1, 1, 1, 1, 512, 64, 1, 
Attn_Kmap,1, 1, 1, 1, 512, 64, 1, 
Attn_Qmap,1, 1, 1, 1, 512, 64, 1, 
Attn_QxK', 1, 64, 1, 1, 1, 64, 1,
Attn_SFMXxV, 1, 64, 1, 1, 64, 1, 1,

The first 3 corresponding to the mapping and split of the head. No issues there.
'Attn_QxK', 1, 64, 1, 1, 1, 64, 1' is where I think the problem is.
If we follow the rule of convolution, the output would be of size 1x64x64. While it should really be of 1 x length x length, and 1x1x1 in this case since the length is 1.

More details:
I believe this line of the configuration file corresponds part (only one head) of the computation of this [line](https://github.com/tensorflow/models/blob/c80de2ca3ec34c4d510961b6a604309772fc02ad/official/transformer/model/attention_layer.py#L129) in the official transformer model:
logits = tf.matmul(q, k, transpose_b=True)

q & k are results of the split_heads function, which is of shape [batch_size, num_heads, length, depth].(tf.reshape(x, [batch_size, length, self.num_heads, depth]).transpose([0, 2, 1, 3]) from line 66&69)

[batch_size, num_heads, length_x, depth] * [batch_size, num_heads, length_y, depth]T should be [batch_size, num_heads, length_x, length_y], and [1, 1, 1] in the case of this configuration file since there is no batching considering, the length is 1 and this config file part only considers one head.

And the line 'Attn_SFMXxV, 1, 64, 1, 1, 64, 1, 1' is thus also problematic.

**Comment by aqeelanwar** [comment]
Hi Zinechant

I have had the same confusion and few other confusion but wasnt able to verify. Here is what I think it should have been. 

<img width="715" alt="image" src="https://user-images.githubusercontent.com/29272667/60147790-99e5e080-9783-11e9-8dc4-ed8d2ee785a3.png">


1. Sentence length is 1
2. The embedding layer basically multiplies the input one-hot (1x1x33708) vector with a learned embedding-matrix to generate the 1x1x512 vector. Its like a dense layer except there is no bias. Hence the channel is equal to the dimension of input vector, and Num filter equals the output dimension.

2. Attn_QxK' should be a scalar . i.e. multiplication of 1x64 dimensional vector with 64x1 dimensional vector. Hence  Its again like a dense layer with input 1x1x64 and output 1x1x1. Here the input feature map is the vector Q and the weight being multiplied with is vector K'

3. Softmax(QK')*V takes input the scalar and multiples it with the vector V. Its dense formulation will be going from input 1x1x1 to 1x1x64. QK' can be taken as the input feature map and the weights to be V. This dense layer will then generate the output in the required way

4. The concatenation step shouldnt be there after every head calculation. Rather once all the heads have been calculated, we need to carry out a multiplication between the concatenated output of all heads (1x512) with a matrix that generates the output 1x64. Again dense layer representaion.


I am not sure if this approach is right. What do you think?

**Comment by whatmough** [comment]
We'll take this discussion over to v2, here: https://github.com/scalesim-project/scale-sim-v2
